services/wireguard/wireguard.go

```diff
@@ -0,0 +1,136 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package wireguard
+
+import (
+	"encoding/json"
+	"net"
+
+	"github.com/mysteriumnetwork/node/money"
+	dto_discovery "github.com/mysteriumnetwork/node/service_discovery/dto"
+)
+
+// ServiceType indicates "wireguard" service type
+const ServiceType = "wireguard"
+
+// ServiceDefinition structure represents "wireguard" service parameters
+type ServiceDefinition struct {
+	// Approximate information on location where the service is provided from
+	Location dto_discovery.Location `json:"location"`
+}
+
+// GetLocation returns geographic location of service definition provider
+func (service ServiceDefinition) GetLocation() dto_discovery.Location {
+	return service.Location
+}
+
+// PaymentMethod indicates payment method for Wireguard service
+const PaymentMethod = "WG"
+
+// Payment structure describes price for Wireguard service payment
+type Payment struct {
+	Price money.Money `json:"price"`
+}
+
+// GetPrice returns price of payment per time
+func (method Payment) GetPrice() money.Money {
+	return method.Price
+}
+
+// ConnectionEndpoint represents Wireguard network instance, it provide information
+// required for establishing connection between service provider and consumer.
+type ConnectionEndpoint interface {
+	Start(config *ServiceConfig) error
+	AddPeer(publicKey string, endpoint *net.UDPAddr) error
+	Config() (ServiceConfig, error)
+	Stop() error
+}
+
+// ServiceConfig represent a Wireguard service provider configuration that will be passed to the consumer for establishing a connection.
+type ServiceConfig struct {
+	Provider struct {
+		PublicKey string
+		Endpoint  net.UDPAddr
+	}
+	Consumer struct {
+		PrivateKey string // TODO peer private key should be generated on consumer side
+	}
+	Subnet net.IPNet
+}
+
+// MarshalJSON implements json.Marshaler interface to provide human readable configuration.
+func (s ServiceConfig) MarshalJSON() ([]byte, error) {
+	type provider struct {
+		PublicKey string `json:"public_key"`
+		Endpoint  string `json:"endpoint"`
+	}
+	type consumer struct {
+		PrivateKey string `json:"private_key"`
+	}
+
+	return json.Marshal(&struct {
+		Provider provider `json:"provider"`
+		Consumer consumer `json:"consumer"`
+		Subnet   string   `json:"subnet"`
+	}{
+		provider{
+			s.Provider.PublicKey,
+			s.Provider.Endpoint.String(),
+		},
+		consumer{
+			s.Consumer.PrivateKey,
+		},
+		s.Subnet.String(),
+	})
+}
+
+// UnmarshalJSON implements json.Unmarshaler interface to receive human readable configuration.
+func (s *ServiceConfig) UnmarshalJSON(data []byte) error {
```

Comment:
I am not sure we still this complicated unmarshaller - maybe just that config struct bellow could be made exported and simply used?

Reply:
The only purpose of this Unmarshaler is to return a human-readable config.
```JSON
{
	"provider": {
		"public_key": "hw253QllKd/qRBdLfc5hfDgyEO6woz6XZf2++N2TNzM=",
		"endpoint": "18.206.227.146:52820"
	},
	"consumer": {
		"private_key": "iMjn+gBtvWWjFaKtTdVizEX0hVBv4YCKKg9IHKKSIV8="
	},
	"subnet": "10.182.47.1/24"
}
```
instead of 
```JSON
{
	"Provider": {
		"PublicKey": "/p6EsyHjMC8kofoXbdIWdtdc/xaVqtsIP5sLPNyklAs=",
		"Endpoint": {
			"IP": "18.206.227.146",
			"Port": 52820,
			"Zone": ""
		}
	},
	"Consumer": {
		"PrivateKey": "IO3+92cVNy0OZqnfyDik3cUtSu3WWaNfa2RVqGgaLXY="
	},
	"Subnet": {
		"IP": "10.182.47.1",
		"Mask": "////AA=="
	}
}
```

Reply:
And I'd prefer to keep `net.UDPAddr` and `net.IPNet` in the struct instead of `string` to keep convetion only in marshaler/unmashaler, but not in every place where I need it.

Reply:
Sounds reasonable. Let's see how it works